File: src/WN_S.cpp
#include <iostream>
#include <fstream>
#include <regex>

using namespace std;

string* idiomas = new string[5] { "cat", "eus", "glg","por","spa"};

string do_replace( string const & in, string const & from, string const & to )
{
  return regex_replace( in, regex(from), to );
}

int contar_synsets(int *synsets, int s, size_t size)
{
    int count = 1;

    auto itr = find(synsets, synsets + size, s);

    if(itr == synsets + size)
    {
        return count;
    }

    for (size_t i = distance(synsets, itr); i < size; i++)
    {
        if(synsets[i] == s) count++;
    }
    
    return count;   
}


string codificacion_categorias(string c)
{
    if(c == "n") return "1";
    else if (c == "v") return "2";
    else if (c == "a") return "3";
    else if (c == "r") return "4";
    else return "-1";
}

int main()
{   
    for(int idiomaIndex = 0; idiomaIndex < 5; idiomaIndex++)
    {
        string datos [7];
        int synsets[150000];
        fstream ficheroLectura;
        fstream ficheroEscritura;
        size_t i = 0;
        size_t j = 0;

        cout << "Abriendo ficheros...\n";
        
        ficheroLectura.open("../mcr/"+idiomas[idiomaIndex]+"WN/wei_"+idiomas[idiomaIndex]+"-30_variant.tsv", ios::in);    
        ficheroEscritura.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_s.pl", ios::out);
        
        if(!ficheroLectura.is_open() || !ficheroEscritura.is_open())
        {
            cout << "No se han abierto correctamente\n";
            return 0;
        }

        cout << "Se han abierto correctamente.\n" << "Parseando a Prolog...\n";

        string line;

        while(getline(ficheroLectura, line))
        {
            i = 0;
            char *cadena = line.data();
            char *token =strtok(cadena, "\t");

            while(token != NULL)
            {
                datos[i] = token;
                token = strtok(NULL, "\t");
                i++;
            }

            string synsetAux = codificacion_categorias(datos[3]) + datos[2].substr(7, 8);
            int synset = stoi(synsetAux);
            ficheroEscritura << "s(" << synset << "," << contar_synsets(synsets, synset, j) << ",'" << do_replace(datos[0], "'", "''") << "'," << datos[3] << "," << datos[1] << ",'')." << endl;
            synsets[j] = synset;
            j++;
        }

        cout << "Idioma "+idiomas[idiomaIndex]+" parseado correctamente.\n";
        ficheroLectura.close();
        ficheroEscritura.close();
    }
    cout << "¡Proceso finalizado!";
    return 0;
}

File: src/WN_S.py
import pandas as pd
import time

idiomas = ['cat', 'eng', 'eus', 'glg', 'por', 'spa']

def codificacionCategoria(c):
    categorias = {
        'n':'1',
        'v':'2',
        'a':'3',
        'r':'4'
        }
    return categorias[c]

def crearSynsetID(synset):
    synset = codificacionCategoria(synset[-1]) + synset[7:15]
    return int(synset)

def contarSynsets(lista, e):
   contador = 0
   for i in lista:
      if i == e:
         contador += 1
   return str(contador)

def ponerEspacios(palabra):
    if type(palabra) == str:
        return "\'" + palabra[1:-1].replace('\'','\'\'').replace('_',' ') + "\'"

for i in idiomas:
    inicio = time.time()

    ruta = "mcrCSV//" + i + "WN\wei_" + i + "-30_variant.csv"
    print("Leyendo dataframe " + ruta)
    
    df = pd.read_csv(ruta, index_col = [0])
    df = df.drop(columns = ['Conf', 'Exp', 'Mark'])
    df.columns = ['Word', 'Sense', 'Synset', 'Type']

    df['Synset'] = df['Synset'].apply(crearSynsetID)
    df['Word'] = df["Word"].apply(ponerEspacios)

    W_Nums = []
    Synsets_list = df.Synset.value_counts()

    # TODO: se podría optimizar un poco más? Se puede utilizar apply o transform al revés? :D
    # Tiempos de la última ejecución:
    # cat ha tardado 16.22.
    # eng ha tardado 39.88.
    # eus ha tardado  7.83.
    # glg ha tardado  8.45.
    # por ha tardado  4.94.
    # spa ha tardado 25.50.

    for index in reversed(df.index):
        fila = df.loc[index]
        W_Nums.insert(0, Synsets_list[fila['Synset']])
        Synsets_list[fila['Synset']] = Synsets_list[fila['Synset']] - 1

    df['W Num'] = W_Nums
    df['Tag Count'] = [0] * len(df['Word'])

    df = df[['Synset', 'W Num', 'Word', 'Type', 'Sense', 'Tag Count']]

    df = df.sort_values(by = ['Synset','W Num'])
    df = df.reset_index(drop = True)
    df.to_csv(i + "\PrologCSV\wn_s.csv")

    ficheroEscritura = open(i + "\Prolog\wn_s.pl", "w", encoding = 'utf-8')

    for i in df.index:
        ficheroEscritura.write("s(" + str(df["Synset"][i]) + "," + str(df["W Num"][i]) + "," + str(df["Word"][i]) + "," + df["Type"][i] + "," + str(df["Sense"][i]) + "," + str(df["Tag Count"][i]) + ").\n")

    ficheroEscritura.close()

    final = time.time()
    print('Proceso en ruta: ' + ruta + ' finalizado. Ha tardado ' + str(final-inicio) + '.\n')
